sys-utils/flock.1
.\" -----------------------------------------------------------------------
.\"
.\"   Copyright 2003-2006 H. Peter Anvin - All Rights Reserved
.\"
.\"   Permission is hereby granted, free of charge, to any person
.\"   obtaining a copy of this software and associated documentation
.\"   files (the "Software"), to deal in the Software without
.\"   restriction, including without limitation the rights to use,
.\"   copy, modify, merge, publish, distribute, sublicense, and/or
.\"   sell copies of the Software, and to permit persons to whom
.\"   the Software is furnished to do so, subject to the following
.\"   conditions:
.\"
.\"   The above copyright notice and this permission notice shall
.\"   be included in all copies or substantial portions of the Software.
.\"
.\"   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
.\"   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
.\"   OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
.\"   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
.\"   HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
.\"   WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
.\"   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\"   OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" -----------------------------------------------------------------------
.TH FLOCK 1 "July 2014" "util-linux" "User Commands"
.SH NAME
flock \- manage locks from shell scripts
.SH SYNOPSIS
.B flock
[options]
.IR file | "directory command " [ arguments ]
.br
.B flock
[options]
.IR file | directory
.BI \-c " command"
.br
.B flock
.RI [options] " number"
.SH DESCRIPTION
.PP
This utility manages
.BR flock (2)
locks from within shell scripts or from the command line.
.PP
The first and second of the above forms wrap the lock around the execution of a
.IR command ,
in a manner similar to
.BR su (1)
or
.BR newgrp (1).
They lock a specified \fIfile\fR or \fIdirectory\fR, which is created (assuming
appropriate permissions) if it does not already exist.  By default, if the
lock cannot be immediately acquired,
.B flock
waits until the lock is available.
.PP
The third form uses an open file by its file descriptor \fInumber\fR.
See the examples below for how that can be used.
.SH OPTIONS
.TP
.BR \-c , " \-\-command " \fIcommand
Pass a single \fIcommand\fR, without arguments, to the shell with
.BR \-c .
.TP
.BR \-E , " \-\-conflict\-exit\-code " \fInumber
The exit code used when the \fB\-n\fP option is in use, and the
conflicting lock exists, or the \fB\-w\fP option is in use,
and the timeout is reached.  The default value is \fB1\fR.
.TP
.BR \-e , " \-x" , " \-\-exclusive"
Obtain an exclusive lock, sometimes called a write lock.  This is the
default.
.TP
.BR \-n , " \-\-nb" , " \-\-nonblock"
Fail rather than wait if the lock cannot be
immediately acquired.
See the
.B \-E
option for the exit code used.
.TP
.BR \-o , " \-\-close"
Close the file descriptor on which the lock is held before executing
.IR command .
This is useful if
.I command
spawns a child process which should not be holding the lock.
.TP
.BR \-s , " \-\-shared"
Obtain a shared lock, sometimes called a read lock.
.TP
.BR \-u , " \-\-unlock"
Drop a lock.  This is usually not required, since a lock is automatically
dropped when the file is closed.  However, it may be required in special
cases, for example if the enclosed command group may have forked a background
process which should not be holding the lock.
.TP
.BR \-w , " \-\-wait" , " \-\-timeout " \fIseconds
Fail if the lock cannot be acquired within
.IR seconds .
Decimal fractional values are allowed.
See the
.B \-E
option for the exit code used. The zero number of
.IR seconds
is interpreted as \fB\-\-nonblock\fR.
.TP
.B \-\-verbose
Tell why lock could not be got, and when it can be how long did it take
to acquire the lock.
.TP
.BR \-V , " \-\-version"
Display version information and exit.
.TP
.BR \-h , " \-\-help"
Display help text and exit.
.SH EXAMPLES
.TP
shell1> flock /tmp -c cat
.TQ
shell2> flock -w .007 /tmp -c echo; /bin/echo $?
Set exclusive lock to directory /tmp and the second command will fail.
.TP
shell1> flock -s /tmp -c cat
.TQ
shell2> flock -s -w .007 /tmp -c echo; /bin/echo $?
Set shared lock to directory /tmp and the second command will not fail.
Notice that attempting to get exclusive lock with second command would fail.
.TP
shell> flock -x local-lock-file echo 'a b c'
Grab the exclusive lock "local-lock-file" before running echo with 'a b c'.
.TP
(
.TQ
  flock -n 9 || exit 1
.TQ
  # ... commands executed under lock ...
.TQ
) 9>/var/lock/mylockfile
The form is convenient inside shell scripts.  The mode used to open the file
doesn't matter to
.BR flock ;
using
.I >
or
.I >>
allows the lockfile to be created if it does not already exist, however,
write permission is required.  Using
.I <
requires that the file already exists but only read permission is required.
.TP
[ "${FLOCKER}" != "$0" ] && exec env FLOCKER="$0" flock -en "$0" "$0" "$@" || :
This is useful boilerplate code for shell scripts.  Put it at the top of the
shell script you want to lock and it'll automatically lock itself on the first
run.  If the env var $FLOCKER is not set to the shell script that is being run,
then execute flock and grab an exclusive non-blocking lock (using the script
itself as the lock file) before re-execing itself with the right arguments.  It
also sets the FLOCKER env var to the right value so it doesn't run again.
.SH "EXIT STATUS"
The command uses
.B sysexits.h
return values for everything, except when using either of the options
.B \-n
or
.B \-w
which report a failure to acquire the lock with a return value given by the
.B \-E
option, or 1 by default.
.PP
When using the \fIcommand\fR variant, and executing the child worked, then
the exit status is that of the child command.
.SH AUTHOR
.UR hpa@zytor.com
H. Peter Anvin
.UE
.SH COPYRIGHT
Copyright \(co 2003\-2006 H. Peter Anvin.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.BR flock (2)
.SH AVAILABILITY
The flock command is part of the util-linux package and is available from
.UR ftp://\:ftp.kernel.org\:/pub\:/linux\:/utils\:/util-linux/
Linux Kernel Archive
.UE .


sys-utils/flock.c
/*   Copyright 2003-2005 H. Peter Anvin - All Rights Reserved
 *
 *   Permission is hereby granted, free of charge, to any person
 *   obtaining a copy of this software and associated documentation
 *   files (the "Software"), to deal in the Software without
 *   restriction, including without limitation the rights to use,
 *   copy, modify, merge, publish, distribute, sublicense, and/or
 *   sell copies of the Software, and to permit persons to whom
 *   the Software is furnished to do so, subject to the following
 *   conditions:
 *
 *   The above copyright notice and this permission notice shall
 *   be included in all copies or substantial portions of the Software.
 *
 *   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 *   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 *   OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 *   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 *   HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 *   WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 *   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 *   OTHER DEALINGS IN THE SOFTWARE.
 */

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <paths.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#include "c.h"
#include "nls.h"
#include "strutils.h"
#include "closestream.h"
#include "monotonic.h"

static void __attribute__((__noreturn__)) usage(int ex)
{
	fprintf(stderr, USAGE_HEADER);
	fprintf(stderr,
		_(" %1$s [options] <file>|<directory> <command> [<argument>...]\n"
		  " %1$s [options] <file>|<directory> -c <command>\n"
		  " %1$s [options] <file descriptor number>\n"),
		program_invocation_short_name);

	fputs(USAGE_SEPARATOR, stderr);
	fputs(_("Manage file locks from shell scripts.\n"), stderr);

	fputs(USAGE_OPTIONS, stderr);
	fputs(_(  " -s, --shared             get a shared lock\n"), stderr);
	fputs(_(  " -x, --exclusive          get an exclusive lock (default)\n"), stderr);
	fputs(_(  " -u, --unlock             remove a lock\n"), stderr);
	fputs(_(  " -n, --nonblock           fail rather than wait\n"), stderr);
	fputs(_(  " -w, --timeout <secs>     wait for a limited amount of time\n"), stderr);
	fputs(_(  " -E, --conflict-exit-code <number>  exit code after conflict or timeout\n"), stderr);
	fputs(_(  " -o, --close              close file descriptor before running command\n"), stderr);
	fputs(_(  " -c, --command <command>  run a single command string through the shell\n"), stderr);
	fputs(_(  "     --verbose            increase verbosity\n"), stderr);
	fprintf(stderr, USAGE_SEPARATOR);
	fprintf(stderr, USAGE_HELP);
	fprintf(stderr, USAGE_VERSION);
	fprintf(stderr, USAGE_MAN_TAIL("flock(1)"));
	exit(ex);
}

static sig_atomic_t timeout_expired = 0;

static void timeout_handler(int sig __attribute__((__unused__)),
			    siginfo_t *info,
			    void *context __attribute__((__unused__)))
{
	if (info->si_code == SI_TIMER)
		timeout_expired = 1;
}

static int open_file(const char *filename, int *flags)
{

	int fd;
	int fl = *flags == 0 ? O_RDONLY : *flags;

	errno = 0;
	fl |= O_NOCTTY | O_CREAT;
	fd = open(filename, fl, 0666);

	/* Linux doesn't like O_CREAT on a directory, even though it
	 * should be a no-op; POSIX doesn't allow O_RDWR or O_WRONLY
	 */
	if (fd < 0 && errno == EISDIR) {
		fl = O_RDONLY | O_NOCTTY;
		fd = open(filename, fl);
	}
	if (fd < 0) {
		warn(_("cannot open lock file %s"), filename);
		if (errno == ENOMEM || errno == EMFILE || errno == ENFILE)
			exit(EX_OSERR);
		if (errno == EROFS || errno == ENOSPC)
			exit(EX_CANTCREAT);
		exit(EX_NOINPUT);
	}
	*flags = fl;
	return fd;
}

static int setup_timer(timer_t *t_id, struct itimerval *timeout)
{
	struct sigaction sig_a;
	static struct sigevent sig_e = {
		.sigev_notify = SIGEV_SIGNAL,
		.sigev_signo = SIGALRM
	};
	struct itimerspec val = {
		.it_value.tv_sec = timeout->it_value.tv_sec,
		.it_value.tv_nsec = timeout->it_value.tv_usec * 1000,
		.it_interval.tv_sec = 0,
		.it_interval.tv_nsec = 0
	};

	if (sigemptyset(&sig_a.sa_mask))
		return 1;
	sig_a.sa_flags = SA_SIGINFO;
	sig_a.sa_handler = (void (*)(int))timeout_handler;
	if (sigaction(SIGALRM, &sig_a, 0))
		return 1;
	if (timer_create(CLOCK_MONOTONIC, &sig_e, t_id))
		return 1;
	if (timer_settime(*t_id, SA_SIGINFO, &val, NULL))
		return 1;
	return 0;
}

static void cancel_timer(timer_t *t_id)
{
	timer_delete(*t_id);
}

int main(int argc, char *argv[])
{
	static timer_t t_id;
	struct itimerval timeout;
	int have_timeout = 0;
	int type = LOCK_EX;
	int block = 0;
	int open_flags = 0;
	int fd = -1;
	int opt, ix;
	int do_close = 0;
	int status;
	int verbose = 0;
	struct timeval time_start, time_done;
	/*
	 * The default exit code for lock conflict or timeout
	 * is specified in man flock.1
	 */
	int conflict_exit_code = 1;
	char **cmd_argv = NULL, *sh_c_argv[4];
	const char *filename = NULL;
	enum {
		OPT_VERBOSE = CHAR_MAX + 1
	};
	static const struct option long_options[] = {
		{"shared", no_argument, NULL, 's'},
		{"exclusive", no_argument, NULL, 'x'},
		{"unlock", no_argument, NULL, 'u'},
		{"nonblocking", no_argument, NULL, 'n'},
		{"nb", no_argument, NULL, 'n'},
		{"timeout", required_argument, NULL, 'w'},
		{"wait", required_argument, NULL, 'w'},
		{"conflict-exit-code", required_argument, NULL, 'E'},
		{"close", no_argument, NULL, 'o'},
		{"verbose", no_argument, NULL, OPT_VERBOSE},
		{"help", no_argument, NULL, 'h'},
		{"version", required_argument, NULL, 'V'},
		{NULL, 0, NULL, 0}
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	atexit(close_stdout);

	if (argc < 2)
		usage(EX_USAGE);

	memset(&timeout, 0, sizeof timeout);

	optopt = 0;
	while ((opt =
		getopt_long(argc, argv, "+sexnouw:E:hV?", long_options,
			    &ix)) != EOF) {
		switch (opt) {
		case 's':
			type = LOCK_SH;
			break;
		case 'e':
		case 'x':
			type = LOCK_EX;
			break;
		case 'u':
			type = LOCK_UN;
			break;
		case 'o':
			do_close = 1;
			break;
		case 'n':
			block = LOCK_NB;
			break;
		case 'w':
			have_timeout = 1;
			strtotimeval_or_err(optarg, &timeout.it_value,
				_("invalid timeout value"));
			break;
		case 'E':
			conflict_exit_code = strtos32_or_err(optarg,
				_("invalid exit code"));
			break;
		case OPT_VERBOSE:
			verbose = 1;
			break;
		case 'V':
			printf(UTIL_LINUX_VERSION);
			exit(EX_OK);
		default:
			/* optopt will be set if this was an unrecognized
			 * option, i.e.  *not* 'h' or '?
			 */
			usage(optopt ? EX_USAGE : 0);
			break;
		}
	}

	if (argc > optind + 1) {
		/* Run command */
		if (!strcmp(argv[optind + 1], "-c") ||
		    !strcmp(argv[optind + 1], "--command")) {
			if (argc != optind + 3)
				errx(EX_USAGE,
				     _("%s requires exactly one command argument"),
				     argv[optind + 1]);
			cmd_argv = sh_c_argv;
			cmd_argv[0] = getenv("SHELL");
			if (!cmd_argv[0] || !*cmd_argv[0])
				cmd_argv[0] = _PATH_BSHELL;
			cmd_argv[1] = "-c";
			cmd_argv[2] = argv[optind + 2];
			cmd_argv[3] = 0;
		} else {
			cmd_argv = &argv[optind + 1];
		}

		filename = argv[optind];
		fd = open_file(filename, &open_flags);

	} else if (optind < argc) {
		/* Use provided file descriptor */
		fd = strtos32_or_err(argv[optind], _("bad file descriptor"));
	} else {
		/* Bad options */
		errx(EX_USAGE, _("requires file descriptor, file or directory"));
	}

	if (have_timeout) {
		if (timeout.it_value.tv_sec == 0 &&
		    timeout.it_value.tv_usec == 0) {
			/* -w 0 is equivalent to -n; this has to be
			 * special-cased because setting an itimer to zero
			 * means disabled!
			 */
			have_timeout = 0;
			block = LOCK_NB;
		} else
			if (setup_timer(&t_id, &timeout))
				err(EX_OSERR, _("cannot not setup timer"));
	}

	if (verbose)
		gettime_monotonic(&time_start);
	while (flock(fd, type | block)) {
		switch (errno) {
		case EWOULDBLOCK:
			/* -n option set and failed to lock. */
			if (verbose)
				warnx(_("failed to get lock"));
			exit(conflict_exit_code);
		case EINTR:
			/* Signal received */
			if (timeout_expired) {
				/* -w option set and failed to lock. */
				if (verbose)
					warnx(_("timeout while waiting to get lock"));
				exit(conflict_exit_code);
			}
			/* otherwise try again */
			continue;
		case EIO:
		case EBADF:		/* since Linux 3.4 (commit 55725513) */
			/* Probably NFSv4 where flock() is emulated by fcntl().
			 * Let's try to reopen in read-write mode.
			 */
			if (!(open_flags & O_RDWR) &&
			    type != LOCK_SH &&
			    filename &&
			    access(filename, R_OK | W_OK) == 0) {

				close(fd);
				open_flags = O_RDWR;
				fd = open_file(filename, &open_flags);

				if (open_flags & O_RDWR)
					break;
			}
			/* go through */
		default:
			/* Other errors */
			if (filename)
				warn("%s", filename);
			else
				warn("%d", fd);
			exit((errno == ENOLCK
			      || errno == ENOMEM) ? EX_OSERR : EX_DATAERR);
		}
	}

	if (have_timeout)
		cancel_timer(&t_id);
	if (verbose) {
		struct timeval delta;

		gettime_monotonic(&time_done);
		timersub(&time_done, &time_start, &delta);
		printf(_("%s: getting lock took %ld.%06ld seconds\n"),
		       program_invocation_short_name, delta.tv_sec,
		       delta.tv_usec);
	}
	status = EX_OK;

	if (cmd_argv) {
		pid_t w, f;
		/* Clear any inherited settings */
		signal(SIGCHLD, SIG_DFL);
		if (verbose)
			printf(_("%s: executing %s\n"), program_invocation_short_name, cmd_argv[2]);
		f = fork();

		if (f < 0) {
			err(EX_OSERR, _("fork failed"));
		} else if (f == 0) {
			if (do_close)
				close(fd);
			execvp(cmd_argv[0], cmd_argv);
			/* execvp() failed */
			warn(_("failed to execute %s"), cmd_argv[0]);
			_exit((errno == ENOMEM) ? EX_OSERR : EX_UNAVAILABLE);
		} else {
			do {
				w = waitpid(f, &status, 0);
				if (w == -1 && errno != EINTR)
					break;
			} while (w != f);

			if (w == -1) {
				status = EXIT_FAILURE;
				warn(_("waitpid failed"));
			} else if (WIFEXITED(status))
				status = WEXITSTATUS(status);
			else if (WIFSIGNALED(status))
				status = WTERMSIG(status) + 128;
			else
				/* WTF? */
				status = EX_OSERR;
		}
	}

	return status;
}
